src/test_snmp.cpp

```diff
@@ -538,3 +539,78 @@ TEST_F(SNMPTest, SuccessFailCountByRequestTypeTable)
   cwtest_reset_time();
   delete tbl;
 }
+
+// Advance to the next start of interval - accurate to within the first
+// second.
+void jump_to_next_periodstart(uint32_t interval_ms)
+{
+  struct timespec now;
+  clock_gettime(CLOCK_REALTIME_COARSE, &now);
+  uint64_t ms_since_epoch = (now.tv_sec * 1000) + (now.tv_nsec / 1000000);
+  cwtest_advance_time_ms(interval_ms - (ms_since_epoch % interval_ms));
+}
+
+TEST_F(SNMPTest, ContinuousAccumulatorTable)
+{
+  cwtest_completely_control_time();
+
+  // Consider a 5 minute period
+  jump_to_next_periodstart(300000);
+
+  // Advance to 2m into the period
+  cwtest_advance_time_ms(120000);
+
+  // Create table at this point
+  SNMP::ContinuousAccumulatorTable* tbl = SNMP::ContinuousAccumulatorTable::create("continuous", test_oid);
+
+  // Add one value to the table and advance 90 seconds
+  tbl->accumulate(200);
+  cwtest_advance_time_ms(90000);
+
+  // Average should not take into account the time before the table was created
+  // (therefore average should still be 200)
+  FILE* fd = popen("snmpwalk -v2c -On -c clearwater 127.0.0.1:16161 .1.2.2.1.2.2", "r");
+  char buf[1024];
+  fgets(buf, sizeof(buf), fd);
+  ASSERT_STREQ(".1.2.2.1.2.2 = Gauge32: 200\n", buf); // Current 5 minutes
+
+  // Add another value to the table and advance 89 seconds
+  tbl->accumulate(100);
+  cwtest_advance_time_ms(89000);
+
+  // This period has spent half the time at 200, and half at 100
+  // So average value should be 150
+  fd = popen("snmpwalk -v2c -On -c clearwater 127.0.0.1:16161 .1.2.2.1.2.2", "r");
+  fgets(buf, sizeof(buf), fd);
+  ASSERT_STREQ(".1.2.2.1.2.2 = Gauge32: 150\n", buf); // Current 5 minutes
+
+  // Jump forward to the next 5 minutes (half way through it)
+  cwtest_advance_time_ms(150000);
+
+  // The average value should be 100 for the current 5 minutes as it is carried
+  // over, additionally, should be value for HWM and LWM, and variance 0
+  fd = popen("snmpwalk -v2c -On -c clearwater 127.0.0.1:16161 .1.2.2.1.2.2", "r");
+  fgets(buf, sizeof(buf), fd); // Average
+  ASSERT_STREQ(".1.2.2.1.2.2 = Gauge32: 100\n", buf); // Current 5 minutes
+
+  fd = popen("snmpwalk -v2c -On -c clearwater 127.0.0.1:16161 .1.2.2.1.3.2", "r");
+  fgets(buf, sizeof(buf), fd); // Variance
+  ASSERT_STREQ(".1.2.2.1.3.2 = Gauge32: 0\n", buf); // Current 5 minutes
+
+  fd = popen("snmpwalk -v2c -On -c clearwater 127.0.0.1:16161 .1.2.2.1.4.2", "r");
+  fgets(buf, sizeof(buf), fd); // HWM
+  ASSERT_STREQ(".1.2.2.1.4.2 = Gauge32: 100\n", buf); // Current 5 minutes
+
+  fd = popen("snmpwalk -v2c -On -c clearwater 127.0.0.1:16161 .1.2.2.1.5.2", "r");
+  fgets(buf, sizeof(buf), fd); // LWM
+  ASSERT_STREQ(".1.2.2.1.5.2 = Gauge32: 100\n", buf); // Current 5 minutes
+
+  // The previous 5 minutes should not have changed value
+  fd = popen("snmpwalk -v2c -On -c clearwater 127.0.0.1:16161 .1.2.2.1.2.3", "r");
+  fgets(buf, sizeof(buf), fd);
+  ASSERT_STREQ(".1.2.2.1.2.3 = Gauge32: 150\n", buf); // Current 5 minutes
+
```

Comment:
Worth checking variance, LWM and HWM on previous period (or before incrementing) - we don't currently check that, for example, we've got LWM and HWM the right way round.